[ieee802154] mux + new userspace interface

bbbert opened the pull request:
Adds a multiplexer and userspace interface for IEEE 802.15.4 transmission/reception and configuration.

- [x] Multiplex
- [x] Trait interface for key / device management
- [x] Fixed-size list implementation of key/device management
- [x] Replace stubs in MAC for key/device lookup
- [x] Add note about Thread's fixed extended address
- [x] Move MAC-related things into ieee802154 namespace
- [x] Driver: think through multi-process support
- [x] Driver: mechanism to return data to userspace
- [x] Userspace: Figure out a better way to deal with signed transmission power
- [x] Userspace: headers
- [x] Userspace: representation of keys and addreses
- [x] Userspace: init/config
- [x] Userspace: neighbor management
- [x] Userspace: key management
- [x] Userspace: transmit/receive
- [x] Change format of frame data that is copied to userspace
- [x] Userspace: packet inspection
- [x] Userspace: update examples

niklasad1 commented on userland/libtock/ieee802154.c:
```diff
@@ -0,0 +1,118 @@
+#include "ieee802154.h"
+#include "t"
```
I guess this is a typo?

bbbert replied:
Definitely a typo, along with `res >= 0` which should be `res > 0`.

niklasad1 commented:
So far you got alot of warnings for unused code, is this because all functionality is not implemented yet?
Otherwise you should fix it

However, I have just skimmed through the PR and it looks promising keep it up.

bbbert commented:
This is still very much a WIP, I put this up in case I'm completely barking down the wrong tree with respect to how driver code should look. There's going to be a lot of repetitive glue code, so I'd rather not rewrite things if possible.

Namely, could I get some advice on the userland/kernel driver side of things? Namely, is it a good idea to abuse `SuccessWithValue` by encoding data into positive numbers? I would rather avoid this if possible, but it seems rather wasteful to `allow()` 2-byte or 1-byte buffers just to pass a small amount of information in, when it could easily be packed into < 31 bits. Am I missing a better way to pass data in/out of the kernel?

bbbert commented:
This PR, like #581, messes with the chip layout because the kernel code size exceeded the ROM region previously allotted to it. Expanding this region and decreasing the size of the PROG region makes the code compile. However, by eliminating other changes, I realize that this prevents userland apps from running.

So far, I found one other place (`userland/tools/flash/imix.sh`) where the PROG offset needs to be changed, but it still doesn't work. Is there anything else (hardcoded constants, etc) that needs to be done, or is it just impossible to increase the size of the ROM region?

ppannuto commented:
Hmm.. PROG_LENGTH is also set here: https://github.com/helena-project/tock/blob/master/userland/userland_generic.ld#L8 , but since it's at a fixed sentinel offset ( https://github.com/helena-project/tock/blob/master/userland/userland_generic.ld#L18 ), I would be a little surprised if making it smaller were an issue, I don't think any of our apps are near the size limit

alevy commented:
I doubt PROG_LENGTH is an issue, but it might make sense that the offset makes a difference.

alevy commented on boards/imix/chip_layout.ld:
```diff
@@ -1,8 +1,8 @@
-PROG_LENGTH = 0x00040000;
+ROM_LENGTH  = 0x00025000;
+PROG_ORIGIN = 0x00035000;
+PROG_LENGTH = 0x00035000;
```
I think this 448K flash limit is a holder from storm's bootloader, and is already accounted for otherwise. We have 512kB total.

To simplify things, the layout should probably be:

```
ROM_ORIGIN  = 0x00010000;
ROM_LENGTH  = 0x00030000;
PROG_ORIGIN = 0x00040000;
PROG_LENGTH = 0x00040000;
```

bbbert replied:
Got it, thanks.

alevy commented:
@bbbert is it correct that this PR includes only a 802.15.4 layer userland interface, and that an IP layer interface is to come?

Since this is WIP, it _might_ be easier to break this up into a PR with just the virtualization layer (which seems pretty reasonable) and another with the userland PR, which also seems reasonable but might end up encountering some feedback on the specific interface (it'll also make merging into #559 a bit easier for me later :P).

bbbert commented on boards/imix/src/main.rs:
```diff
@@ -61,8 +61,7 @@ struct Imix {
     ninedof: &'static capsules::ninedof::NineDof<'static>,
-    radio: &'static capsules::radio::RadioDriver<'static,
-                                                 capsules::mac::MacDevice<'static, RF233Device>>,
+    radio_driver: &'static capsules::ieee802154::RadioDriver<'static>,
```
I thought about that too, and then I noticed that all the other user-facing `Driver`s seem to be suffixed by `Driver`. Also, given that it is namespaced within `ieee802154`, I think `RadioDriver` is actually appropriate. Other radios would then be `ble::RadioDriver` or `ieee80211::RadioDriver` or whatever. What do you think?

alevy replied:
sure... why not just `Driver` or `SyscallDriver` then if we're relying on the package name for the actual meaning? (i.e. would there be an ieee802154 non-Radio driver?). the HIL still ought to have a different name, but fine... I don't care that much about the name...

bbbert commented:
Found the issue with the program offsets: tockloader itself is still flashing to the old address. Adding a `--app-address` flag to `tockloader install` lets apps run again. I'm testing the rest of the userland functionality and will update this PR soon.

@alevy Yep, I don't think there's a point to writing an IP interface over a fragmentation interface that's not yet stabilized, a pending ieee802154, and a pending aes-ccm implementation over a pending encryption interface. It would just lead to rebase hell.

As for breaking out virtualization, I would have to modify the old radio driver (I moved the `Mac` trait and changed initialization to store a trait reference instead of a reference to a trait-bounded type)
and then later replace it completely, which seems a bit wasteful too. I thought it might be cleaner to just replace everything at once.

bbbert commented:
@ppanuto Thanks for letting me know about that other occurrence of `PROG_LENGTH`. I don't think it caused any harm and definitely should not anymore since I'm now keeping `PROG_LENGTH` unchanged.
@alevy Userspace now seems to be working, so I guess this PR is ready for review.

bradjc commented on capsules/src/ieee802154/driver.rs:
```diff
@@ -0,0 +1,827 @@
+//! Implements a userspace interface for sending and receiving IEEE 802.15.4
```
This should have a short description, then a blank comment line, then this paragraph to show up better on `rustdoc`.

bbbert replied:
Cool, will do.

bradjc commented on capsules/src/ieee802154/driver.rs:
```diff
@@ -0,0 +1,827 @@
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Return radio status. SUCCESS/EOFF = on/off.
```
Tock has a convention that calling command 0 should always return >= 0 if the driver is present (i.e. the kernel has support for it) and < 0 if the kernel isn't compiled with that `Driver`. So here, you can either define a return of `1` as on and `0` as off (so that whether the radio is on or off userland can still detect that there is a 15.4 driver present) or you can move each command up a number and add a new command 0 that just returns `SUCCESS` when called (as many other capsule drivers do).

phil-levis commented:
I talked with Hubert a little bit about this yesterday.

I think it makes sense to have a few different make options for sizing. E.g., ‘make’ for a “standard” app/kernel sizing, ‘make big-kernel’ for more kernel space and perhaps ‘make small-kernel’ for more app space. Whether or not there’s a small, medium, or big kernel shouldn’t be tied to particular modules. 

In the best case, kernel compilation should figure out the size of the kernel and generate a proper layout for it.

Phil

> On Aug 28, 2017, at 4:12 PM, Hubert Teo <notifications@github.com> wrote:
> 
> This PR, like #581, messes with the chip layout because the kernel code size exceeded the ROM region previously allotted to it. Expanding this region and decreasing the size of the PROG region makes the code compile. However, by eliminating other changes, I realize that this prevents userland apps from running.
> 
> So far, I found one other place (userland/tools/flash/imix.sh) where the PROG offset needs to be changed, but it still doesn't work. Is there anything else (hardcoded constants, etc) that needs to be done, or is it just impossible to increase the size of the ROM region?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

alevy commented:
@phil-levis yeah that's something we should be able to do pretty easily. It's out of scope for this PR though, and I think the new sizing is a reasonable general default given how little flash we've been using so far for processes, so I think it's fine to fix this in a different PR.

alevy commented:
There are two comments that @bradjc left that still need to be addressed: one is about module comment formatting, the other for driver subcommand numbering. @bbbert are you gonna be able to address those still?

I'm happy merging after that.

bbbert commented:
@alevy Sorry, I was somewhere remote for the past four days. Actually, I think those two comments have already been addressed.

I'd still like to fix the driver naming though, as per your suggestion. I can do that in about two days when I am back at my laptop. Otherwise, merging this and renaming in a separate PR is also fine with me.
